### DISCUSSION TITLE
Correct use of new "interpolate"

### DISCUSSION THREAD

**Discussion by andrewlhicks** [firedrakeproject/firedrake#3674] [Firedrake support] [answered]
I'm currently trying to convert my code from the old way of interpolating to the new way, i.e. to
```
from firedrake.__future__ import interpolate

f = assemble(interpolate(some_expr, some_func_space))
```

While simply doing the old method, i.e.
```
from firedrake import interpolate

f = interpolate(some_expr, some_func_space)
```
gives a warning, I noticed that simply writing
```
f = Function(some_func_space)
f.interpolate(some_expr)
```
doesn't throw a warning. Is interpolating via the second method (i.e. using the `interpolate` _method_ as opposed to the freestanding `interpolate` _function_) still going to be supported then? I'm asking because I still have lots of code that uses this syntax to interpolate stuff. I'm not sure if I need to rewrite my code in this case.

**Comment by connorjward** [comment]
The `interpolate` _method_ of `Function` (as opposed to the `interpolate` _function_) is always evaluated eagerly so you don't need to change anything in that case.

**Comment by dham** [comment, marked as answer]
The difference is that `interpolate` the free function now returns a symbol for interpolation, which can be composed with other symbolic operations and evaluated later. That's the change we needed to make to increase composability. `Function.interpolate` never returned anything and isn't needed as a symbolic operator, so it doesn't change.

**Reply by andrewlhicks** [reply]
Thanks for your answer. So just to clarify, the two methods of interpolation, i.e.
```
f = Function(some_func_space)
f.interpolate(some_expr)
```
and
```
f = assemble(interpolate(some_expr, some_func_space))
```
give the same result?

**Reply by dham** [reply]
Yes, that's correct.
